server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1296,6 +1382,29 @@ private BlobStoreIndexShardSnapshot loadShardSnapshot(BlobContainer shardContain
         }
     }
 
+    /**
+     * Loads all available snapshots in the repository using the given {@code generation} or falling back to trying to determine it from
+     * the given list of blobs in the shard container.
+     *
+     * @param blobs      list of blobs in repository
+     * @param generation shard generation or {@code null} in case there was no shard generation tracked in the {@link RepositoryData} for
+     *                   this shard because its snapshot was created in a version older than
+     *                   {@link SnapshotsService#SHARD_GEN_IN_REPO_DATA_VERSION}.
+     * @return tuple of BlobStoreIndexShardSnapshots and the last snapshot index generation
+     */
+    private Tuple<BlobStoreIndexShardSnapshots, String> buildBlobStoreIndexShardSnapshots(Set<String> blobs,
+                                                                                          BlobContainer shardContainer,
+                                                                                          @Nullable String generation) throws IOException {
+        if (generation != null) {
+            if (generation.equals(ShardGenerations.NEW_SHARD_GEN)) {
```

Comment:
I would prefer  to move this logic to the caller (there is just one where this is applicable) and assert in this method that the generation is not one of these special generation types

Reply:
See the previous comments on `NEW_SHARD_GEN` persistency -> There's two where it's applicable. The delete can also see this when there's only a failed snapshot (e.g. where the node left and never updated master) for the shard.

Reply:
ok